### PR TITLE
chore(sdk): swap out sharedpreferences

### PR DESCRIPTION
**Pull request by anonvt**
Introduces the EventPipeline, in which events are collected and persisted asynchronously (with relation the main thread), and batched for upload.

Closes:
https://topsort.atlassian.net/browse/API-797
https://topsort.atlassian.net/browse/API-800

PR: https://github.com/Topsort/topsort.kt/pull/11

### CODE REVIEW

File: TopsortAnalytics/src/androidTest/java/com/topsort/analytics/EventPipelineTest.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        val impressions2 = listOf(
+            getImpressionPromoted(),
+            getImpressionOrganic()
+        )
+
+        runBlocking {
+            EventPipeline.storeImpression(ImpressionEvent(impressions1))
+            EventPipeline.storeImpression(ImpressionEvent(impressions2))
+            // Make sure they're persisted
+            delay(10)
```
Comment (fcs-ts) [inline TopsortAnalytics/src/androidTest/java/com/topsort/analytics/EventPipelineTest.kt]: why do we need this delay? note that when we start running tests on these repo, these delays will add up to our test runtime

File: TopsortAnalytics/src/main/java/com/topsort/analytics/EventPipeline.kt [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            }
+        }
+
+    private suspend fun aggregateEvents(): String {
+        val data = applicationContext.eventDatastore.data.first()
+        val impressions = data[KEY_IMPRESSION_EVENTS]?.trim(',')
+        val clicks = data[KEY_CLICK_EVENTS]?.trim(',')
+        val purchases = data[KEY_PURCHASE_EVENTS]?.trim(',')
+
+        val aggregated = StringBuilder()
```
Comment (fcs-ts) [inline TopsortAnalytics/src/main/java/com/topsort/analytics/EventPipeline.kt]: Can we just build and object that aggregates the individual events and serialize it? Manually creating JSONs is very error-prone.

Comment (anonvt) [reply TopsortAnalytics/src/main/java/com/topsort/analytics/EventPipeline.kt]: Impressions, clicks and purchases are stored as a serialized list that is incrementally expanded, as a compromise I've deserialized them into each component (ImpressionEvent etc) and aggregated them into an Event.
In the future we might consider using the Protobuf datastore to store these as objects directly to provide better type safety.

File: TopsortAnalytics/src/androidTest/java/com/topsort/analytics/EventPipelineTest.kt (hunk 1 above)
Comment (anonvt) [reply TopsortAnalytics/src/androidTest/java/com/topsort/analytics/EventPipelineTest.kt]: Worked into some joins and yields to not require the delays, which I agree are sad to see.

**Review by fcs-ts — APPROVED** [review]

**Comment by jbergstroem** [comment]
Note: this is still flagged as draft; should we review for land?

**Review by PabloReszczynski — APPROVED** [review]
lgtm

**Pull request merged**